include/mbgl/mtl/context.hpp

```diff
@@ -0,0 +1,157 @@
+#pragma once
+
+#include <mbgl/gfx/texture.hpp>
+#include <mbgl/gfx/draw_mode.hpp>
+#include <mbgl/gfx/depth_mode.hpp>
+#include <mbgl/gfx/stencil_mode.hpp>
+#include <mbgl/gfx/color_mode.hpp>
+#include <mbgl/gfx/texture2d.hpp>
+#include <mbgl/gfx/context.hpp>
+#include <mbgl/mtl/buffer_resource.hpp>
+#include <mbgl/mtl/mtl_fwd.hpp>
+#include <mbgl/util/noncopyable.hpp>
+
+#include <memory>
+#include <unordered_map>
+
+namespace mbgl {
+
+class ProgramParameters;
+
+namespace mtl {
+
+class RendererBackend;
+class ShaderProgram;
+
+using UniqueShaderProgram = std::unique_ptr<ShaderProgram>;
+
+class Context final : public gfx::Context {
+public:
+    Context(RendererBackend&);
+    ~Context() noexcept override;
+    Context(const Context&) = delete;
+    Context& operator=(const Context& other) = delete;
+
+    std::unique_ptr<gfx::CommandEncoder> createCommandEncoder() override;
+
+    gfx::RenderingStats& renderingStats() { return stats; }
+    const gfx::RenderingStats& renderingStats() const override { return stats; }
+
+    BufferResource createBuffer(const void* data, std::size_t size, gfx::BufferUsageType) const;
+
+    UniqueShaderProgram createProgram(std::string name,
+                                      std::string_view source,
+                                      std::string_view vertexName,
+                                      std::string_view fragmentName,
+                                      const ProgramParameters& programParameters,
+                                      const std::unordered_map<std::string, std::string>& additionalDefines);
+
+    MTLTexturePtr createMetalTexture(MTLTextureDescriptorPtr textureDescriptor) const;
+    MTLSamplerStatePtr createMetalSamplerState(MTLSamplerDescriptorPtr samplerDescriptor) const;
+
+    /*
```

Comment:
I am assuming this comment block will be uncommented in a future PR.

Reply:
Yes, the target/framebuffer stuff should be implemented shortly.  I've removed some of that which definitely won't be replaced.